### PR TITLE
fix: import and export ed25519

### PR DESCRIPTION
**Pull request by ThatOneBro**
Fixes import and export, keeps a copy of "seed" for export when key is a private ed25519 key.

### CODE REVIEW

**Comment by ThatOneBro** [comment]
Screen Shot 2023-02-06 at 5 25 39 PM
Seems like all tests are passing now @panva

File: src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp [hunk 1]
```diff
@@ -294,26 +294,38 @@ ExceptionOr<Vector<uint8_t>> CryptoKeyOKP::exportPkcs8() const
...
 
-    return base64URLEncodeToString(Span<const uint8_t> { publicKey, sizeof(publicKey) });
+    return Span<const uint8_t> { public_key, sizeof(public_key) };
 }
 
-String CryptoKeyOKP::getX25519PublicFromPrivate() const
+Span<const uint8_t> CryptoKeyOKP::x25519PublicFromPrivate(KeyMaterial& privateKey)
 {
-    uint8_t publicKey[X25519_PUBLIC_VALUE_LEN];
+    uint8_t public_key[X25519_PUBLIC_VALUE_LEN];
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp]: Isn't this a use-after-stack-free? public_key and private key is allocated on the stack

Comment (Jarred-Sumner) [reply src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp]: Instead, KeyMaterial& would need to be a reference holding the publicKey and privateKey

File: src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp [hunk 2]
```diff
@@ -301,31 +301,33 @@ String CryptoKeyOKP::generateJwkD() const
...
 }
 
-Span<const uint8_t> CryptoKeyOKP::x25519PublicFromPrivate(KeyMaterial& privateKey)
+CryptoKeyOKP::KeyMaterial CryptoKeyOKP::x25519PublicFromPrivate(const KeyMaterial& privateKey)
 {
-    uint8_t public_key[X25519_PUBLIC_VALUE_LEN];
+    auto publicKey = KeyMaterial(X25519_PUBLIC_VALUE_LEN);
 
-    X25519_public_from_private(public_key, privateKey.data());
+    X25519_public_from_private(publicKey.data(), privateKey.data());
```
Comment (ThatOneBro) [inline src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp:318]: Is it bad to do this?

File: src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp [hunk 3]
```diff
@@ -301,31 +301,33 @@ String CryptoKeyOKP::generateJwkD() const
...
 {
-    uint8_t public_key[ED25519_PUBLIC_KEY_LEN], private_key[ED25519_PRIVATE_KEY_LEN];
+    uint8_t publicKey[ED25519_PUBLIC_KEY_LEN];
+    auto privateKey = KeyMaterial(ED25519_PRIVATE_KEY_LEN);
 
-    ED25519_keypair_from_seed(public_key, private_key, seed.data());
+    ED25519_keypair_from_seed(publicKey, privateKey.data(), seed.data());
 
-    return WTFMove(Vector<uint8_t>(private_key, ED25519_PRIVATE_KEY_LEN));
+    return WTFMove(privateKey);
```
Comment (ThatOneBro) [inline src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp:330]: Is it ok to do this? It should avoid copy right?

File: src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp [hunk 4]
```diff
@@ -301,31 +301,33 @@ String CryptoKeyOKP::generateJwkD() const
     return base64URLEncodeToString(m_data);
 }
 
-Span<const uint8_t> CryptoKeyOKP::ed25519PublicFromPrivate(KeyMaterial& privateKey)
+CryptoKeyOKP::KeyMaterial CryptoKeyOKP::ed25519PublicFromPrivate(const KeyMaterial& seed)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp:304]: This should be `CryptoKeyOKP::KeyMaterial&&` I think

File: src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp [hunk 5]
```diff
@@ -301,31 +301,33 @@ String CryptoKeyOKP::generateJwkD() const
...
 
-    ED25519_keypair_from_seed(public_key, private_key, privateKey.data());
+    ED25519_keypair_from_seed(publicKey.data(), privateKey, seed.data());
 
-    return Span<const uint8_t> { public_key, sizeof(public_key) };
+    return WTFMove(publicKey);
 }
 
-Span<const uint8_t> CryptoKeyOKP::x25519PublicFromPrivate(KeyMaterial& privateKey)
+CryptoKeyOKP::KeyMaterial CryptoKeyOKP::x25519PublicFromPrivate(const KeyMaterial& privateKey)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/webcrypto/CryptoKeyOKPOpenSSL.cpp:314]: `CryptoKeyOKP::KeyMaterial&&`

**Pull request merged**